File: fixtures/scale/dart/large/lib/ciphers/porta/porta_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'porta_key.dart';

/// Reciprocal cipher over half-alphabets selected by the key letter.
class PortaCipher implements Cipher {
  final PortaKey key;

  PortaCipher(this.key);

  @override
  String name() => 'porta';

  @override
  String encrypt(String plaintext) => _swapHalves(cleanAlphabet(plaintext));

  @override
  String decrypt(String ciphertext) => _swapHalves(cleanAlphabet(ciphertext));

  String _swapHalves(String text) {
    final sb = StringBuffer();
    for (var i = 0; i < text.length; i++) {
      final x = indexOf(text[i]);
      final row = indexOf(key.keyCharAt(i)) ~/ 2;
      final y = x < 13 ? 13 + (x + row) % 13 : (x - 13 - row) % 13;
      sb.write(charAt(y < 0 ? y + 13 : y));
    }
    return sb.toString();
  }
}

File: fixtures/scale/java/large/src/cryptobench/ciphers/porta/PortaCipher.java
package cryptobench.ciphers.porta;

import cryptobench.core.Cipher;
import cryptobench.util.Alphabet;

/** Reciprocal cipher over half-alphabets selected by the key letter. */
public class PortaCipher implements Cipher {
    private final PortaKey key;

    public PortaCipher(PortaKey key) {
        this.key = key;
    }

    @Override
    public String name() {
        return "porta";
    }

    @Override
    public String encrypt(String plaintext) {
        return swapHalves(Alphabet.clean(plaintext));
    }

    @Override
    public String decrypt(String ciphertext) {
        return swapHalves(Alphabet.clean(ciphertext));
    }

    private String swapHalves(String text) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < text.length(); i++) {
            int x = Alphabet.indexOf(text.charAt(i));
            int row = Alphabet.indexOf(key.keyCharAt(i)) / 2;
            int y = x < 13 ? 13 + Math.floorMod(x + row, 13) : Math.floorMod(x - 13 - row, 13);
            sb.append(Alphabet.charAt(y));
        }
        return sb.toString();
    }
}
